source/Various/Hydra.lagda

```diff
@@ -0,0 +1,182 @@
+
+data HeadLocation₁ : List Hydra → 𝓤₀ ̇  where
+ here₀ : {hs hs' : List Hydra} → HeadLocation₀ hs → HeadLocation₁ (Node hs ∷ hs')
+ here₁ : {hs hs' : List Hydra} → HeadLocation₁ hs → HeadLocation₁ (Node hs ∷ hs')
+ next  : {h : Hydra} {hs : List Hydra} → HeadLocation₁ hs → HeadLocation₁ (h ∷ hs)
+
+HeadLocation : Hydra → 𝓤₀ ̇
+HeadLocation (Node hs) = HeadLocation₀ hs + HeadLocation₁ hs
+
+\end{code}
+
+Similarly, we define two cutting functions, one for cutting child head
+and for cutting a grand-child head and regrowing subheads
+
+\begin{code}
+
+cons-mult : {X : 𝓤 ̇ } → X → ℕ → List X → List X
+cons-mult x 0 xs = xs
+cons-mult x (succ n) xs = x ∷ (cons-mult x n xs)
+
+cut₀ : (hs : List Hydra) → HeadLocation₀ hs → List Hydra
+cut₀ (h ∷ hs) (here)    = hs 
+cut₀ (h ∷ hs) (next l)  = h ∷ (cut₀ hs l)
+
+cut₁ : ℕ → (hs : List Hydra) → HeadLocation₁ hs → List Hydra
+cut₁ n (h ∷ hs')         (next  l) = h ∷ (cut₁ n hs' l)
+cut₁ n ((Node hs) ∷ hs') (here₀ l) = cons-mult (Node (cut₀ hs l)) (succ n) hs'
+cut₁ n ((Node hs) ∷ hs') (here₁ l) = Node (cut₁ n hs l) ∷ hs'
+
+cut : ℕ → (h : Hydra) → HeadLocation h → Hydra
+cut n (Node hs) (inl l) = Node (cut₀ hs l)
+cut n (Node hs) (inr l) = Node (cut₁ n hs l)
+
+\end{code}
+
+To prove that the process of cutting hydra's heads always end, we define
+a relation between hydras and prove it's well founded.
+
+\begin{code}
+
+_⊲_ : Hydra → Hydra → 𝓤₀ ̇
+h ⊲ h' = Σ n ꞉ ℕ , Σ l ꞉ HeadLocation h' , h ＝ cut n h' l
+
+pattern step n l eq = (n , l , eq)
+
+_⊲'_ : List Hydra → List Hydra → 𝓤₀ ̇
+hs ⊲' hs' = Σ n ꞉ ℕ , ( ( Σ l ꞉ HeadLocation₀ hs' , hs ＝ cut₀ hs' l)
+                      + ( Σ l ꞉ HeadLocation₁ hs' , hs ＝ cut₁ n hs' l))
+
+pattern step₀ n l eq = (n , inl (l , eq))
+pattern step₁ n l eq = (n , inr (l , eq))
+
+⊲⇒⊲' : {hs hs' : List Hydra} → Node hs ⊲ Node hs' → hs ⊲' hs'
+⊲⇒⊲' (step n (inl l) refl) = step₀ n l refl
+⊲⇒⊲' (step n (inr l) refl) = step₁ n l refl
+
+[]-is-minimum : (hs : List Hydra) → ¬(hs ⊲' [])
+[]-is-minimum h (step₀ _ () _)
+[]-is-minimum h (step₁ _ () _)
+
+⊲-is-well-founded : is-well-founded _⊲_
+⊲'-is-well-founded : is-well-founded _⊲'_
+
+⊲-is-well-founded (Node hs) = acc (recurs (⊲'-is-well-founded hs))
+ where
+  recurs : {hs : List Hydra}
+         → is-accessible (_⊲'_) hs
+         → (h' : Hydra) → h' ⊲ Node hs
+         → is-accessible (_⊲_) h'
+  recurs (acc rec₁) (Node hs') r = acc (recurs (rec₁ hs' (⊲⇒⊲' r)))
+
+⊲'-is-well-founded [] = acc (λ h r → 𝟘-elim ([]-is-minimum h r))
+⊲'-is-well-founded (h ∷ hs) = acc (recurs (⊲-is-well-founded h) (⊲'-is-well-founded hs))
+ where
+  recurs : {h : Hydra} {hs : List Hydra}
+         → is-accessible _⊲_ h
+         → is-accessible _⊲'_ hs
+         → (hs' : List Hydra) → hs' ⊲' (h ∷ hs)
+         → is-accessible _⊲'_ hs'
+  recurs (acc rec₁) (acc rec₂) hs' (step₀ n here refl)      = acc rec₂
+  recurs (acc rec₁) (acc rec₂) hs' (step₀ n (next l) refl)  = acc (recurs (acc rec₁) (rec₂ _ (step₀ n l refl)))
```

Comment:
I suggest to break this to avoid long lines:
```
recurs (acc rec₁) (acc rec₂) hs' (step₀ n (next l) refl)  = 
 acc (recurs (acc rec₁) (rec₂ _ (step₀ n l refl)))
```
Our convention, not always followed, particularly in the past, is that we should not exceed 80 columns.